rules/TypeDeclaration/ValueObject/AddParamTypeForFunctionLikeWithinCallLikeArgDeclaration.php

```diff
@@ -16,13 +19,14 @@
     /**
      * @param int<0, max>|string $callLikePosition
      * @param int<0, max> $functionLikePosition
+     * @param Type|Closure(Arg[], NodeTypeResolver): ?Type $paramType
      */
     public function __construct(
         private string $className,
         private string $methodName,
         private int|string $callLikePosition,
         private int $functionLikePosition,
-        private Type $paramType
+        private Type|Closure $paramType
```

Comment:
Could you stick with the `Type` only? Passing closures can go wild very quickly, as any closure can be passed

Reply:
@TomasVotruba I'm not sure how I could make this work without a closure. At best I could implement an interface for an invokable class if that would be better?

Reply:
Looking at the provided code snippet, there is no need for any dynamic code. The type can be inferred from 1st arg:

```diff
-$app->extend(Foo::class, function ($fooInstance) {
+$app->extend(Foo::class, function (Foo $fooInstance) {
});
```

By standard rules should never require dynamic configuration. That should be part of implementation inside `Rector` rule.

Looking at [the rule configuration](https://getrector.com/rule-detail/add-param-type-for-function-like-within-call-like-arg-declaration-rector), this is a different use case.

I think this feature should be turned into a new rule and closure should be moved there.

Reply:
The dynamic use is necessary because there's no way to detect what the type hit should be with the Closure. A better example of the diff would be:

```diff
-$app->extend(Foo::class, function ($fooInstance) {
+$app->extend(Foo::class, function (Foo $fooInstance) {
});
-$app->extend(Bar::class, function ($barInstance) {
+$app->extend(Bar::class, function (Bar $barInstance) {
});
```

I understand not wanting to use Closures in value objects. It felt off to me as well.

I can make additional rules, but there would need to be two rules at least. One for using the static property `class` in an Argument and another for using the type passed as an argument. The code between all three would be 95% identical.

A possible further option is to use a string for these scenarios instead? E.g. you can provide a Type or a string that is `class_string_from_arg:0` or `type_from_arg:1` and the rule can then manage this using a service to determine the type.

Reply:
Could you pop up a separate rule? I still see it as a simple static one.